jupyterkernelgen/jupyterkernelgen.py

```diff
@@ -0,0 +1,181 @@
+        print(f"- {colors.OK}found ipython{colors.ENDC}")
+    else:
+        print(f"- {colors.WARNING}no ipython found{colors.ENDC}")
+    print()
+
+    return found_ipython and found_kernel
+
+def install_ipykernel(conda_exe, conda_env):
+    # only install if the user enters `y`
+    print(f"{colors.HEADER}INSTALL ipykernel? [y/N]{colors.ENDC} ", end="", flush=True)
+
+    inp = sys.stdin.read(1)
+    if inp == "y":
+        try:
+            subprocess.run([conda_exe, "install", "-p", conda_env, "-y", "ipykernel"])
+        except subprocess.CalledProcessError as e:
+            print(f"{colors.FAIL}ipykernel installation failed: {e}{colors.ENDC}", file=sys.stderr)
+            exit(1)
+    else:
+        # exit the script if the user does not want to install ipykernel since
+        # we cannot continue without it
+        print(f"{colors.FAIL}not installing ipykernel. Exiting...{colors.ENDC}", file=sys.stderr)
+        sys.exit(0)
+    print()
+
+def get_kernel_name():
+    print(f"{colors.HEADER}ENTER KERNEL NAME{colors.ENDC}: ", end="", flush=True)
+    kernel_name = sys.stdin.readline().strip()
+    print()
+    return kernel_name
+
+def create_kernel_dir(kernel_name):
+    print(f"{colors.HEADER}CREATING KERNEL DIRECTORY...{colors.ENDC}")
+    try:
+        # replace ~ with the path to the user's home directory
+        path = os.path.expanduser(f"~/.local/share/jupyter/kernels/{kernel_name}")
+        if os.path.exists(path):
+            print(f"{colors.FAIL}Environment with that name already exists. Exiting...{colors.ENDC}", file=sys.stderr)
+            sys.exit(1)
+        os.makedirs(path, exist_ok=True)
+    except OSError as e:
+        print(f"{colors.FAIL}Failed to create directory: {e}{colors.ENDC}", file=sys.stderr)
+        sys.exit(1)
+    print()
+    return path
+
+def create_kernel_helper_script(path, conda_env):
+    print(f"{colors.HEADER}CREATING KERNEL HELPER SCRIPT...{colors.ENDC}")
+    source = f"""#!/bin/bash
+
+source activate {conda_env}
+exec "$@"
+"""
+    try:
+        f = open(path + "/kernel-helper.sh", "w")
+        f.write(source)
+        f.close()
+    except OSError as e:
+        print(f"{colors.FAIL}failed to write to {path}/kernel-helper.sh: {e}{colors.ENDC}", file=sys.stderr)
+        sys.exit(1)
+    print()
+
+def create_kernel_json(path, kernel_name):
+    print(f"{colors.HEADER}CREATING KERNEL JSON...{colors.ENDC}")
+    source = f"""{{
+  "argv": ["{{resource_dir}}/kernel-helper.sh", "python3", "-m", "ipykernel_launcher", "-f", "{{connection_file}}"],
+  "display_name": "{kernel_name}",
+  "language": "python"
+}}"""
+    try:
+        f = open(path + "/kernel.json", "w")
+        f.write(source)
+        f.close()
+    except OSError as e:
+        print(f"{colors.FAIL}failed to write to {path}/kernel.json: {e}{colors.ENDC}", file=sys.stderr)
+        sys.exit(1)
+    print()
+
+def main():
+    try:
+        kernel_name = get_kernel_name()
+        path = create_kernel_dir(kernel_name)
```

Comment:
possible to get one new line between all function calls and comments as well.
Normally most people go to the main() function when first looking at what your code is doing. 
More comments in this block allows developer to know what/where things are being done.